Feature recommendation

flyinghermit opened the pull request:
This PR adds:
- `RecommendFeature` type which holds feature recommendation status `'NOTIFY' | 'DONE'` for given feature.
   ```
     export type RecommendFeature = {
        TrustedDevices: RecommendationStatus;
     };
    
     export type RecommendationStatus = 'NOTIFY' | 'DONE';
    ```

- adds a local storage item `grv_recommend_feature` to track `RecommendFeature`.
- shows "red dot" in side navigation if notification status for given feature is `'NOTIFY'` and cluster already has at least one resource connected, which is checked with `const onboard = localStorage.getOnboardDiscover(); onboard.hasResource = True`
- Side navigation menu is selected based on navigation item title. To avoid cases where title is updated in the feature and the string comparison breaks, I have added `enum NavTitles` and updated `TeleportFeatureNavigationItem`
    ```
    TeleportFeatureNavigationItem {
       title: NavTitles;
       ....
   }
     ``` 
- currently supports  trusted devices but the mechanism is designed to be used for other features as well (since we have ongoing plan to highlight other existing enterprise features)

See `e` counterpart for more description - https://github.com/gravitational/teleport.e/pull/1912

flyinghermit committed:
red dot notificaiton

flyinghermit committed:
recommendFeature type and show alert based on configured recommended features

flyinghermit committed:
register recommendFeature, only show if requiresOnboarding returns false

flyinghermit committed:
removed pulse based on design feedback

flyinghermit committed:
use localstorage to persist state:
- setup feature recommendation state with 'Pending' and 'Done'
- state is only updated during context setup for simplicity

flyinghermit committed:
Navivation item Title types

flyinghermit committed:
udpate status type name, fix test

flyinghermit committed:
Merge branch 'master' into sshah/feature-recommendation

flyinghermit committed:
show red dot on onboard?.hasResource

flyinghermit committed:
add RECOMMEND_FEATURE to KEEP_LOCALSTORAGE_KEYS_ON_LOGOUT

sshahcodes commented:
"red dot" in side nav:
<img width="181" alt="Screenshot 2023-08-01 at 7 18 44 PM" src="https://github.com/gravitational/teleport/assets/10177602/31048197-df5a-4e36-b037-5496637f970d">

ryanclark commented on web/packages/teleport/src/types.ts:
```diff
@@ -29,13 +29,59 @@ export interface Context {
   isExternalLink?: boolean;
 }
 
+export type NavTitle = `${NavTitles}`;
```
What's this used for?

sshahcodes replied:
Thanks for pointing out, I defined it to be used as a type built from enum but ended up not using it anywhere. Removed in https://github.com/gravitational/teleport/commit/007fcdb19d5d45a2e664b0aa9208dbe9b446e9ee

ryanclark commented on web/packages/teleport/src/types.ts:
```diff
@@ -118,3 +164,10 @@ export type LockedFeatures = {
+  TrustedDevices: RecommendationStatus;
+};
+
+export type RecommendationStatus = 'NOTIFY' | 'DONE';
```
Can this be an enum?

sshahcodes replied:
Updated this to enum in https://github.com/gravitational/teleport/pull/29757/commits/007fcdb19d5d45a2e664b0aa9208dbe9b446e9ee

flyinghermit committed:
apply review suggestions:
- Update enum name from NavTitle(s) to NavTitle
- remove unused NavTitle type
- make RecommendationStatus an enum

kimlisa commented on web/packages/teleport/src/Navigation/NavigationItem.tsx:
```diff
@@ -154,6 +158,26 @@ export function NavigationItem(props: NavigationItemProps) {
   );
 
+  // showRedDot returns red dot component if the feature recommendation state is 'NOTIFY'
+  function showRedDot(featureName: NavTitle): JSX.Element {
```
`showRedDot` sounds like a flag
```suggestion
  function renderHighlightFeature(featureName: NavTitle): JSX.Element {
```
or something like that

flyinghermit committed:
Merge branch 'master' of github.com:gravitational/teleport into sshah/feature-recommendation

flyinghermit committed:
run prettier

flyinghermit committed:
rename showRedDot -> renderHighlightFeature

flyinghermit committed:
Merge branch 'master' into sshah/feature-recommendation

flyinghermit committed:
match submodule with master
